File: Python/clbg/spectral-norm/perfarena_generated.py
import sys
import math

def solve():
    """
    Reads N from command line arguments and estimates the spectral norm 
    of the matrix A using the power method for N iterations.
    """
    if len(sys.argv) != 2:
        # Should not happen based on problem description, but good practice.
        return

    try:
        N = int(sys.argv[1])
    except ValueError:
        return

    if N <= 0:
        # Handle non-positive N if necessary, though problem states positive integer.
        print(f"{0.0:.9f}")
        return

    # Determine the dimension D. Since indices i and j go up to D-1, 
    # and the process runs for N steps, we assume D = N + 1 is required 
    # to support the indices involved in the calculation.
    D = N + 1

    # Initialize the vector v to a unit vector (e.g., v[0] = 1.0)
    v = [0.0] * D
    v[0] = 1.0

    def get_A_ij(i, j):
        """Calculates A[i][j] = 1 / ((i+j)(i+j+1)/2 + i + 1)"""
        k = i + j
        # T_k = k * (k + 1) / 2.0
        # Denominator = T_k + i + 1.0
        # Using floating point arithmetic throughout.
        denominator = (k * (k + 1.0) / 2.0) + i + 1.0
        return 1.0 / denominator

    def mat_vec_mult_A(vec):
        """Calculates w = A * vec (w_i = sum_j A[i][j] * vec[j])"""
        w = [0.0] * D
        for i in range(D):
            sum_val = 0.0
            for j in range(D):
                A_ij = get_A_ij(i, j)
                sum_val += A_ij * vec[j]
            w[i] = sum_val
        return w

    def mat_vec_mult_AT(vec):
        """Calculates v_new = A^T * vec (v'_j = sum_i A[i][j] * vec[i])"""
        v_new = [0.0] * D
        for j in range(D):
            sum_val = 0.0
            for i in range(D):
                # A^T[j][i] = A[i][j]
                A_ij = get_A_ij(i, j)
                sum_val += A_ij * vec[i]
            v_new[j] = sum_val
        return v_new

    # Power Iteration Loop
    # v_0 is the initial unit vector.
    # Step k=1: v_1 = A * v_0
    # Step k=2: v_2 = A^T * v_1
    # Step k=3: v_3 = A * v_2
    # ...
    # v_N is the final vector.
    
    current_v = list(v) # Start with v_0

    for k in range(1, N + 1):
        if k % 2 == 1:
            # Odd step: v_k = A * v_{k-1}
            current_v = mat_vec_mult_A(current_v)
        else:
            # Even step: v_k = A^T * v_{k-1}
            current_v = mat_vec_mult_AT(current_v)

    # The estimate of the spectral norm is the L2 norm of the final vector v_N.
    norm_sq = sum(x * x for x in current_v)
    estimate = math.sqrt(norm_sq)

    # Output the result formatted to nine decimal places.
    print(f"{estimate:.9f}")

if __name__ == "__main__":
    solve()

File: JavaScript/clbg/spectral-norm/perfarena_generated.js
/**
 * Spectral Norm Benchmark Implementation (Power Iteration)
 * Estimates the spectral norm of a fixed matrix A using N iterations.
 * 
 * Matrix A[i][j] = 1 / ((i+j) * ((i+j+1)/2) + i + 1)
 * Dimension M is assumed to be 1000 based on typical benchmark constraints.
 */

function solve() {
    // Read N from command line arguments
    const N_str = process.argv[2];
    if (!N_str || isNaN(parseInt(N_str))) {
        // Default to 5500 if no valid argument is provided, matching reference default.
        const N = 5500;
        process.stdout.write(String(N));
        return;
    }
    const N = parseInt(N_str);

    // --- Constants and Setup ---
    // Based on typical CLB constraints for this problem, M=1000 is used.
    const M = 1000; 

    // A: The matrix A (M x M)
    // We store A explicitly since A^T is just A transposed.
    const A = new Float64Array(M * M);

    // 1. Precalculate Matrix A
    for (let i = 0; i < M; i++) {
        for (let j = 0; j < M; j++) {
            const S = i + j;
            // Denominator: (i+j) * ((i+j+1)/2) + i + 1
            // Note: Since i+j is an integer, (i+j+1)/2 might be X.5, so we use floating point math.
            const denominator = S * ((S + 1) / 2.0) + i + 1.0;
            A[i * M + j] = 1.0 / denominator;
        }
    }

    // Vectors v, w, v_new (all size M)
    const v = new Float64Array(M);
    const w = new Float64Array(M);
    const v_new = new Float64Array(M);

    // Initialize v to a unit vector (v_0 = [1/sqrt(M), 1/sqrt(M), ...])
    const invSqrtM = 1.0 / Math.sqrt(M);
    for (let i = 0; i < M; i++) {
        v[i] = invSqrtM;
    }

    // --- Power Iteration ---
    for (let k = 0; k < N; k++) {
        // Step 1: w = A * v (Matrix-Vector Multiplication)
        // w[i] = Sum_j A[i][j] * v[j]
        for (let i = 0; i < M; i++) {
            let sum = 0.0;
            for (let j = 0; j < M; j++) {
                // A[i][j] is stored at index i * M + j
                sum += A[i * M + j] * v[j];
            }
            w[i] = sum;
        }

        // Step 2: v_new = A^T * w (Matrix-Vector Multiplication)
        // v_new[j] = Sum_i A^T[j][i] * w[i] = Sum_i A[i][j] * w[i]
        for (let j = 0; j < M; j++) {
            let sum = 0.0;
            for (let i = 0; i < M; i++) {
                // A[i][j] is stored at index i * M + j
                sum += A[i * M + j] * w[i];
            }
            v_new[j] = sum;
        }

        // Step 3: Normalize v_new to get the next v
        let norm = 0.0;
        for (let i = 0; i < M; i++) {
            norm += v_new[i] * v_new[i];
        }
        norm = Math.sqrt(norm);

        // Update v and calculate the norm estimate for the next iteration's tracking
        for (let i = 0; i < M; i++) {
            v[i] = v_new[i] / norm;
        }
        
        // The norm of the resulting vector v_new (before normalization) is the estimate 
        // of the spectral radius of A^T A. We track the norm of the *normalized* vector 
        // for the final output, which should converge to 1 if the process is stable, 
        // but the benchmark expects the final calculated norm magnitude.
        // We use the norm calculated *before* normalization as the estimate for the spectral radius.
        // However, since we overwrite v with the normalized vector, we must calculate the norm 
        // of the *last* vector v_new before normalization, which is the estimate for the spectral norm.
        // We store this estimate in a variable to print at the end.
        if (k === N - 1) {
            // The norm calculated before normalization is the best estimate for the spectral norm.
            // We use the norm calculated in this step.
            // Since we overwrite v, we must re-calculate the norm of the final v_new before normalization.
            // Wait, the final estimate is the norm of the last vector *before* normalization, 
            // which is the norm calculated above.
            // We will use the 'norm' calculated in this loop iteration.
            // Since we are inside the loop, we must store the final norm estimate.
            // We will rely on the fact that the final 'norm' calculated in the last iteration is the answer.
        }
    }

    // The final estimate is the norm calculated in the last iteration (k=N-1).
    // Since the loop structure overwrites 'norm' on every iteration, we must re-calculate 
    // the norm of the final v_new (which was used to calculate the final v) if we want 
    // the exact value from the last step.
    
    // Re-run the final norm calculation using the last v_new calculated (which corresponds to k=N-1)
    let final_norm_sq = 0.0;
    for (let i = 0; i < M; i++) {
        final_norm_sq += v_new[i] * v_new[i];
    }
    const final_estimate = Math.sqrt(final_norm_sq);


    // Output the result formatted to nine decimal places.
    process.stdout.write(final_estimate.toFixed(9) + '\n');
}

solve();
